### PR TITLE
Use new Buildkit "Reservation" Endpoint

### PR DESCRIPTION
**Pull request by brandonSc**

### CODE REVIEW

File: buildkitd/buildkitd.go [hunk 1]
```diff
@@ -594,6 +594,26 @@ func checkConnection(ctx context.Context, address string, opts ...client.ClientO
...
 		defer bkClient.Close()
+		err = bkClient.Reserve(ctxTimeout, 5*time.Second)
+		if err != nil {
+			s, ok := status.FromError(errors.Cause(err))
+			// Note that if err == codes.Unimplemented,
+			// we assume an older version of buildkit and continue to call ListWorkers.
+			switch {
+			case !ok:
+				mu.Lock()
+				connErr = errors.New("unexpected error from buildkit reserve endpoint")
```
Comment (vladaionescu) [inline buildkitd/buildkitd.go:605]: I would not hire the original error in this case - we should at least `Wrap` the original error.

File: buildkitd/buildkitd.go [hunk 2]
```diff
@@ -594,6 +594,26 @@ func checkConnection(ctx context.Context, address string, opts ...client.ClientO
...
+				mu.Unlock()
+			case s.Code() == codes.Unavailable:
+				mu.Lock()
+				connErr = errors.New("buildkit is currently shutting down")
```
Comment (vladaionescu) [inline buildkitd/buildkitd.go:609]: Shutting down or starting up? Why not just use the original `err`?

File: buildkitd/buildkitd.go [hunk 3]
```diff
@@ -594,6 +594,26 @@ func checkConnection(ctx context.Context, address string, opts ...client.ClientO
...
+				mu.Unlock()
+			case s.Code() != codes.Unimplemented:
+				mu.Lock()
+				connErr = err
+				mu.Unlock()
+			}
+		}
```
Comment (vladaionescu) [inline buildkitd/buildkitd.go:616]: I would maybe only keep a simple `if` statement, given that we want to do pretty much the same thing in each case (pass on the original error)

Comment (brandonSc) [reply buildkitd/buildkitd.go:607]: Reworked as you described. Single `if` statement + return original error

**Review by vladaionescu — APPROVED** [review]
Looks good - assuming that the buildkit references are updated to point to a sha in `earthly-main` branch before merging.

**Pull request merged**